jctools-benchmarks/src/main/java/org/jctools/jmh/latency/QueueBurstCost.java

```diff
@@ -68,102 +71,145 @@ public void setupQueue()
-        consumer.stopped.await();
+        final int consumerCount = this.consumerCount;
+        for (int i = 0; i < consumerCount; i++)
+        {
+            consumers[i].isRunning = false;
+        }
+        stopped.await();
     }
 
     @TearDown(Level.Trial)
-    public void stopExecutor() throws InterruptedException
+    public void stopExecutor()
     {
         consumerExecutor.shutdown();
     }
 
     @Benchmark
     @CompilerControl(CompilerControl.Mode.DONT_INLINE)
-    public void burstCost(Stop stop)
+    public void burstCost(Event event)
     {
         final int burst = burstSize;
-        final Queue<AbstractEvent> q = this.q;
-        final Go go = GO;
-        stop.lazySet(false);
-        sendBurst(q, burst, go, stop);
-
-        waitForIt(stop);
+        final Queue<Event> q = this.q;
+        event.reset();
+        sendBurst(q, event, burst);
+        event.waitFor(burst);
     }
 
-    private void waitForIt(Stop stop)
+    private static void sendBurst(Queue<Event> q, Event event, int burst)
     {
-        while (!stop.get())
+        for (int i = 0; i < burst; i++)
         {
-            ;
+            while (!q.offer(event))
+            {
+                ;
+            }
+            if (DELAY_PRODUCER != 0)
+            {
+                Blackhole.consumeCPU(DELAY_PRODUCER);
+            }
         }
     }
 
-    private void sendBurst(Queue<AbstractEvent> q, int burst, Go go, Stop stop)
+    //do not extends AtomicLong here: JMH will take care of padding fields for us
+    @State(Scope.Thread)
+    public static class Event
     {
-        for (int i = 0; i < burst - 1; i++)
+        private static final long ARRAY_BASE;
+        private static final int ELEMENT_SHIFT;
+
+        static
         {
-            while (!q.offer(go))
+            final int scale = UnsafeAccess.UNSAFE.arrayIndexScale(long[].class);
+            if (8 == scale)
             {
-                ;
+                ELEMENT_SHIFT = 3;
             }
-            if (DELAY_PRODUCER != 0)
+            else
             {
-                Blackhole.consumeCPU(DELAY_PRODUCER);
+                throw new IllegalStateException("Unexpected long[] element size");
             }
+            // Including the buffer pad in the array base offset
+            ARRAY_BASE = UnsafeAccess.UNSAFE.arrayBaseOffset(long[].class);
         }
 
-        while (!q.offer(stop))
+        private static long calcSequenceOffset(long index)
         {
-            ;
+            return ARRAY_BASE + (index << ELEMENT_SHIFT);
         }
-    }
 
-    abstract static class AbstractEvent extends AtomicBoolean
-    {
-        abstract void handle();
-    }
+        private long[] handledCount = null;
 
-    static class Go extends AbstractEvent
-    {
-        @Override
-        void handle()
+        @Setup
+        public void init(QueueBurstCost state)
         {
-            // do nothing
+            handledCount = new long[state.consumerCount];
```

Comment:
I think it would be far less complex to put the counter on Consumer, which already has padding. The `waitFor` then looks at the counters on the consumers. You can reduce the reader induced coherence overhead by spin waiting on the final event (as before) and then transition to consumer scanning. Further you can batch update the `handled` counter on the Consumer only when the queue is observed empty or relaxedPoll returns null.

Reply:
> I think it would be far less complex to put the counter on Consumer, which already has padding

That's a nice idea, but the counter is not just per-consumer, need to be per-producer (thread) as well

Reply:
"That's a nice idea, but the counter is not just per-consumer, need to be per-producer (thread) as well."
Fair point, in which case, yes you have to pad the array based counter.

"This would work if the last offered event will be the last to be handled, but with multi-consumer is not ensured anymore"
You can't be sure, but it's fair to assume that there's no need to check the counters before that one in handled.

Reply:
> it's fair to assume that there's no need to check the counters before that one in handled

I don't understand this: maybe with "before" I mean something different.
My understanding of "before" is the index/id of consumer on `Event::handledCount`: if the last (sbmitted) event is being handled, only with a single consumer I can trust that the previous events would be hanlded by it in order, saving the needto poll for their completion, but the current algorithm won't do that anyway (because `Event::handledCount::length == consumerCount == 1`, you ends up always polling).
Help me to understand what do you mean, sorry for being slow in this :)

Reply:
"I can trust that the previous events would be handled with offering order just in the single consumer case, saving to poll for theirs completion."
I'm not saying you should trust this as the only indicator, just that it's a good heuristic to reduce the overall noise.

Reply:
Ok, so I suppose that you mean the `before` as "previous handled events by that same consume": in this case, yes I can handle it like that and see how it goes!

OT:
Do you have had any chance to see the results I've posted? They should ideally replace the ones on http://psy-lob-saw.blogspot.com/2015/01/mpmc-multi-multi-queue-vs-clq.html given that now the mc case should be handled correctly. I'm especially curious about your opinion/experience with asymmetric bench results too.

Reply:
I haven't yet addressed the last change (ie to use last event handling), if you have any concern please send a snipped with the idea you propose to speedup the fix,
I need yet to understand the value of 
```
        @CompilerControl(CompilerControl.Mode.DONT_INLINE)
        private void consume(MessagePassingQueue<Event> q, int consumerId)
```
Why isn't good to be inlined? 
By preventing inlining It would artificially slow don't too fast consumers (that has an high chance of finding the q emtpy) due to increased cost of `consume` when no events are present, thus producing noise.

Reply:
WRT benchmark results, I would appreciate it if you made the effort to format those into something that is more easily comparable.

Reply:
WRT the `consume` method inlining, I can't remember. There's good reasons to avoid inlining in infinite loops to avoid OSR compilations, but I can't remember what was the reasons/testing here.

Reply:
Configuration is the same as https://github.com/franz1981/JCTools-wiki/blob/xadd_docs/MpscUnboundedXaddArrayQueue.md#lies-damned-lies-and-benchmarks
*Symmetric tests*
1P-1C:
![image](https://user-images.githubusercontent.com/13125299/69070447-766c4c80-0a28-11ea-97db-4071ae6ca8bc.png)
2P-2C:
![image](https://user-images.githubusercontent.com/13125299/69070286-27beb280-0a28-11ea-8f27-71aafdadd3d6.png)
3P-3C:
![image](https://user-images.githubusercontent.com/13125299/69070313-34dba180-0a28-11ea-995a-e667b9438eba.png)
4P-4C:
![image](https://user-images.githubusercontent.com/13125299/69070342-4624ae00-0a28-11ea-965a-c9746c0fa6a5.png)
5P-5C:
![image](https://user-images.githubusercontent.com/13125299/69070370-5177d980-0a28-11ea-9ca1-1c5d57db898a.png)
6P-6C:
![image](https://user-images.githubusercontent.com/13125299/69070398-5b014180-0a28-11ea-8a28-a1b9c8558c5d.png)

Reply:
8P-4C:
![image](https://user-images.githubusercontent.com/13125299/69072560-2c856580-0a2c-11ea-9a28-cfaa116c1fcc.png)
10P-2C:
![image](https://user-images.githubusercontent.com/13125299/69072646-5179d880-0a2c-11ea-81f9-539a523b6f0e.png)
11P-1C:
![image](https://user-images.githubusercontent.com/13125299/69072684-6191b800-0a2c-11ea-8f63-26e5d7bcdef4.png)

Reply:
@nitsanw Let me know any insights about them and how you think is performing xadd q in your opinion @ https://github.com/JCTools/JCTools/pull/269